tree/tree/src/TTreeCache.cxx

```diff
@@ -1332,18 +1332,20 @@ Bool_t TTreeCache::FillBuffer()
    Long64_t maxReadEntry = minEntry; // If we are stopped before the end of the 2nd pass, this marker will where we need to start next time.
    Int_t nReadPrefRequest = 0;
    auto perfStats = GetTree()->GetPerfStats();
+
+   struct collectionInfo {
+      Int_t fClusterStart{-1}; // First basket belonging to the current cluster
+      Int_t fCurrent{-1};       // Currently visited basket
+      Bool_t fLoadedOnce{kFALSE};
+
+      void Rewind() { fCurrent = (fClusterStart >= 0) ? fClusterStart : 0; }
+   };
+   std::vector<collectionInfo> cursor(fNbranches);
```

Comment:
In this new version, `cursor` is not anymore initialized at the beginning of the loop. Is this a problem? Perhaps the `do {} while` loop could be prepended by a comment explaining what is being looped over.

Reply:
This lack of reset is the crux of the fix.   This loop is a loop over the clusters (for when the cache can hold more than one cluster worth of the selected branches).   If we reset the cursor within the loop, then we have to find the cluster boundary/start every iteration of the loop (hence the O(N^2))

I am adding the doc:
```
   // Main loop to fill the cache, inside each loop we will loop over 
   // all the cached branch and collect the baskets within the 'current'
   // range/cluster.  If there is still space in the cache after that, we 
   // will do another iteration to add one more cluster to the cache.
   // i.e. essentially loop over the clusters.
```